VFS watcher impl

vemoo opened the pull request:
Implements #433 

For now I ended up referencing  my forked `notify` with the fix until it's merged and released.

There are some things left to do, like what to do on `Rescan` or `Error` events, but the rest it's working.

matklad commented on crates/ra_vfs/src/watcher.rs:
```diff
@@ -0,0 +1,96 @@
+                // forward relevant events only
+                .try_for_each(|change| send_change_events(change, &output_sender))
+                .unwrap()
+        });
```
So this burns a thread to just transform `mpsc` channel into crossbeam one? Seems suboptimal, but I doubt we can do anything better.

vemoo replied:
It's also mapping from `notify`s `DebouncedEvent` to `Tasks`s. I think even if the watcher took a crossbeam channel we would still need a thread to do the mapping

vemoo commented on crates/ra_vfs/src/lib.rs:
```diff
@@ -123,12 +128,13 @@ impl Vfs {
                 filter: Box::new(filter),
             };
             res.worker.inp.send(task).unwrap();
+            res.watcher.watch(path).unwrap();
```
I will add error handling and log a warning for errors.

vemoo commented on crates/ra_vfs/src/lib.rs:
```diff
@@ -97,12 +99,15 @@ impl Vfs {
     pub fn new(mut roots: Vec<PathBuf>) -> (Vfs, Vec<VfsRoot>) {
         let (worker, worker_handle) = io::start();
 
+        let watcher = Watcher::start(worker.inp.clone()).unwrap(); // TODO return Result?
```
Here we're just starting the watcher, there are no paths yet. The paths are added later. The errors that can happen here are from initializing 'inotify' (or platform specific system for watching files) and `mio::Poll` from taking a look at the code. But I see there are [errors](http://man7.org/linux/man-pages/man2/inotify_init.2.html#ERRORS) that are probably worth handling, so I will implement your suggestion.

vemoo commented:
I was missing the `Task:HandleChange` that lets the `Vfs` decide which changes should be loaded. And it was trying to load everything.

After fixing that I'm still seing high CPU usage and the ra_lsp_server sometimes hangs and doesn't close after closing the editor. I will investigate, but I think it's not ready to merge yet.

matklad commented:
Note that I've had to rewrite the history recently, so expect storms while rebasing: https://github.com/rust-analyzer/rust-analyzer/pull/566#issuecomment-455768512

Sorry about that: it's definitively my mistake :slightly_frowning_face:

vemoo commented:
Didn't run into any problems when rebasing.

I found out what the issue was. Using the default recursive watching ment also watching the `target` folder, and during builds there's a lot of activity there. It was so bad that after running `cargo build` and then `cargo clean` my system started swapping after `ra_lsp_server` took all the memory.

I've changed the watching to non recursive and implemented recursive watching taking `.gitignore` into account, using the `ignore` crate. I'm not sure if that's the right way, since that wouldn't work on projects that are not on git. Maybe excluding build folders would be better, but I don't know if there's an easy way to do it.

matklad commented:
>So right know even for changes in the .git folder or node_modules we do that

I'd just hard-code those for now as well :) 

> Maybe the .gitignore support can be implemented in a separate PR?

Makes sense to me! I'll also be fine with just hard-coding worst offenders and punting on this until notify 5.0

>Still, I wonder if it's worth it to store the roots in a hierarchy to avoid always iterating on all of them.

Sure, long-term we need faster data structures everywhere. Everything in rust-analyzer currently uses a dumb linear scan and recomputed every time. However, I think that before starting with "obvious" optimizations, we need to put some integration profiling infrastructure in place. That is, I'd like to see info like "we've spent 5 CPU seconds on traversing the roots in the last minute of rust-analyzer usage".

matklad commented on crates/ra_vfs/src/io/mod.rs:
```diff
@@ -0,0 +1,233 @@
+use std::{
```
We use 2018 style, so `io.rs` instead of `io/mod.rs`

vemoo replied:
I didn't know that, but I like it more.

vemoo commented:
Ok, I've hardcoded `.git` and `node_modules` in any path component aswell.

> Sure, long-term we need faster data structures everywhere. Everything in rust-analyzer currently uses a dumb linear scan and recomputed every time. However, I think that before starting with "obvious" optimizations, we need to put some integration profiling infrastructure in place. That is, I'd like to see info like "we've spent 5 CPU seconds on traversing the roots in the last minute of rust-analyzer usage".

Ok, that makes sense. But I think it's related to how ignoring files should be implemented. Using this project as an example there's a `.gitignore` file at the root level that affets all of the crates. With the current structure all of the `RootFilter`s should know about it. But if there's some hierarchy only the outer root should know about it.

What I'm thinking is a tree with nodes like:
```rust
struct Node {
    ignores: Vec<Ignore>,
    path: PathBuf,
    root: Option<VfsRoot>,
    children: Vec<Node>,
}
```
`root` is optional to be able to create nodes for common paths like cargo registry or the other ignore files are.

JeanMertz commented on crates/ra_vfs/src/lib.rs:
```diff
@@ -16,52 +16,61 @@
+    if path.is_dir() {
+        for (i, c) in rel_path.components().enumerate() {
+            if let Component::Normal(c) = c {
+                // hardcoded for now
```
Perhaps add a `TODO: ` in front of this comment, to make it easier to find things that require some rethinking later?

vemoo replied:
good idea

matklad commented on crates/ra_vfs/src/io.rs:
```diff
@@ -1,55 +1,184 @@
+        root: VfsRoot,
+        path: PathBuf,
+        root_filter: Arc<RootFilter>,
+        nested_roots: Vec<PathBuf>,
```
I think `nested_roots` should just be a part of `RootFilter`, with a more generic name like `excluded_dirs`.

matklad commented on crates/ra_vfs/src/io.rs:
```diff
@@ -1,55 +1,184 @@
+        }
+        Task::HandleChange(change) => {
+            // forward as is because Vfs has to decide if we should load it
+            TaskResult::HandleChange(change)
```
Hm, this looks like one elaborate control flow! I wonder if we could simplify this... At least when I originally thought about it, I've thought of `io` as a separate actor which completely encapsulates file system access and gives very "clean" `TaskResult`s outside:

```rust
enum TaskResult {
    BulkLoadRoot { root: VfsRoot, files: Vec<(Path, String)> },
    AddSingleFile { root: VfsRoot, path: Path, text: String},
    ChangeSingleFile { ... },
    RemoveSingleFile { ... },
}
```

That is, `io` handles all loading and watching of roots,  and  vfs itself then decides if it should ignore `ChangeSingleFile` due to memap or not. Watcher becomes and implementation detail of IO.

I think that probably `Worker` will need to know the set of all roots for this to work, so perhaps we should encapsulate `roots: Arena<VfsRoot, Arc<RootFilter>>,` into `Roots` struct and share it via an `Arc` between vfs and Worker?

Obviously, I don't know the specifics of `notify` API very well, so perhaps this wont' work for some reason?

vemoo replied:
Ok, yes, that sounds better and I think it should work, I will try to implement it tomorrow.

vemoo replied:
Should `Roots` struct be in a `RwLock`? At the moment it can be implemented without it but if the vfs will support adding roots after initialization it will be needed.

matklad replied:
>Should Roots struct be in a RwLock? At the moment it can be implemented without it but if the vfs will support adding roots after initialization it will be needed.

I'd lean towards sharing an immutable snapshot for starters.

vemoo replied:
This is now implemented.

I had to move the watcher to inside the io workers thread:
https://github.com/rust-analyzer/rust-analyzer/blob/1692a11d09cc3d70dcde10cc896c6f7ebb9eb3c3/crates/ra_vfs/src/io.rs#L50-L51
to have access to the output sender and be able to send the `TaskResult`s directly from the watcher.

I've added `excluded_dirs` to `RootFilter` and also
https://github.com/rust-analyzer/rust-analyzer/blob/1692a11d09cc3d70dcde10cc896c6f7ebb9eb3c3/crates/ra_vfs/src/lib.rs#L63
to make using it with `walkdir` cleaner.

The rest is pretty much as you described it, which is a lot cleaner.

One thing I have found is that `WalkDir` by default doesn't follow symlinks which means that `entry.file_type().is_[dir|file]()` and `entry.path().is_[dir|file]()` won't necessarily match. I'm thinking we should probably follow symlinks but i'm not sure.

matklad replied:
Yeah, this now looks superb! I’ll do a manual smoke test tomorrow and merge!

I think it’s ok not to follow simlnks for now

vemoo commented on crates/ra_vfs/src/io.rs:
```diff
@@ -1,55 +1,184 @@
+    AddRoot(AddRootResult),
+    HandleChange(WatcherChange),
+    LoadChange(WatcherChangeData),
+    NoOp,
```
you're right, fixes

matklad commented:
Hm, I still get occasional 100% CPU hangs. Need to investigate this

vemoo commented:
I've found something that may be related https://github.com/passcod/notify/issues/177
